### PR TITLE
Instrument akka dispatchers without relying on Executor/Runnable instrumentations

### PR DESCRIPTION
**Pull request by richardstartin**
* `akka.dispatch.Mailbox` is a subclass of `ForkJoinTask` so `Dispatcher` context propagation should come for free from the j.u.c FJP/FJT instrumentation.  This is almost the case (`akka.dispatch.forkjoin` is needed in older akka versions, and older Play framework versions depend on this).
* `akka.dispatch.ExecutionContexts$sameThreadExecutionContext$` doesn't need context propagation.
* `akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinPool` seems to be FJP based too. This is almost the case: quirky error handling means the _tasks_ but not the executor need special instrumentation, which is guarded by a class loader matcher to prevent field injection of `Runnable` unless this instrumentation is activated (requires Scala 2.11 and an old Akka version so not very likely).
* `akka.dispatch.Dispatcher$LazyExecutorServiceDelegate` is a wrapper around another `ExecutorService`
* I couldn't locate `akka.actor.ActorSystemImpl$$anon$1` but I suspect that not interfering with this is what `akka-init` is testing.

### CODE REVIEW

**Review by mcculls — APPROVED** [review]
LGTM

**Review by bantonsson — APPROVED** [review]
👍

I have some questions with regards how this works compared to the old code. I can't remember if  we tried to propagate things into akka `Dispatcher`s or if things just happened to sneak in there by accident. The thing is though that the `Mailbox` will only get _scheduled_ (i.e. sent to the executer) when it is currently not processing messages, otherwise the message will only be enqueued to the internal queue of the `Mailbox`. This means that `context` will propagate into the actors in a seemingly random way.
Probably not something for this PR to fix, but it would be interesting to know if we've changed the behavior in any significant way from the old instrumentation.

**Comment by richardstartin** [comment]
@bantonsson this was relying on the executor/runnable instrumentation - so only happening when the `Mailbox` gets scheduled. I had been considering instrumenting `Envelope` or possibly `TaskInvocation`.

**Comment by bantonsson** [comment]
Ok, so basically the same behavior as before, but more controlled. I would go for `Envelope` if we want to do proper per message send context propagation , but I'm not sure we want that until we decide how to handle _infinite_ message loops.

**Comment by richardstartin** [comment]
Exactly - the purpose if this PR is just to change this instrumentation such that its behaviour doesn't change when I change things it was depending on.

File: dd-java-agent/instrumentation/akka-concurrent/src/main/java/datadog/trace/instrumentation/akka/concurrent/AkkaForkJoinExecutorTaskInstrumentation.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+  public ElementMatcher<ClassLoader> classLoaderMatcher() {
+    // prevents Runnable from being instrumented unless this
+    // instrumentation would take effect (unless something else
+    // instruments it).
+    return hasClassesNamed("akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask");
+  }
+
+  @Override
+  public Map<String, String> contextStore() {
+    return Collections.singletonMap(Runnable.class.getName(), State.class.getName());
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/akka-concurrent/src/main/java/datadog/trace/instrumentation/akka/concurrent/AkkaForkJoinExecutorTaskInstrumentation.java:46]: Is this expected to stay like this long term?

File: dd-java-agent/instrumentation/akka-concurrent/src/main/java/datadog/trace/instrumentation/akka/concurrent/AkkaForkJoinExecutorTaskInstrumentation.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  @Override
+  public Map<? extends ElementMatcher<? super MethodDescription>, String> transformers() {
+    return singletonMap(
+        isConstructor().and(takesArgument(0, named(Runnable.class.getName()))),
+        getClass().getName() + "$Construct");
+  }
+
+  public static final class Construct {
+    @Advice.OnMethodExit
+    public static void construct(@Advice.Argument(0) Runnable wrapped) {
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/akka-concurrent/src/main/java/datadog/trace/instrumentation/akka/concurrent/AkkaForkJoinExecutorTaskInstrumentation.java:63]: `wrapped`?

File: dd-java-agent/instrumentation/akka-concurrent/src/main/java/datadog/trace/instrumentation/akka/concurrent/AkkaWrappedRunnableInstrumentation.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  }
+
+  @Override
+  public Map<String, String> contextStore() {
+    return Collections.singletonMap(Runnable.class.getName(), State.class.getName());
+  }
+
+  @Override
+  public ElementMatcher<? super TypeDescription> typeMatcher() {
+    return named("akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask");
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/akka-concurrent/src/main/java/datadog/trace/instrumentation/akka/concurrent/AkkaWrappedRunnableInstrumentation.java:52]: both this and `AkkaForkJoinExecutorTaskInstrumentation` are instrumenting the same class.  Why separate them?

File: dd-java-agent/instrumentation/akka-concurrent/src/main/java/datadog/trace/instrumentation/akka/concurrent/AkkaForkJoinTaskInstrumentation.java [hunk 4]
```diff
@@ -41,19 +43,20 @@ public AkkaForkJoinTaskInstrumentation() {
...
   public ElementMatcher<? super TypeDescription> typeMatcher() {
-    return named("akka.dispatch.forkjoin.ForkJoinTask");
+    return extendsClass(named("akka.dispatch.forkjoin.ForkJoinTask"))
+        .and(not(named("akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask")));
   }
 
   @Override
   public Map<? extends ElementMatcher<? super MethodDescription>, String> transformers() {
     Map<ElementMatcher<MethodDescription>, String> transformers = new HashMap<>(4);
-    transformers.put(isMethod().and(named("doExec")), getClass().getName() + "$DoExec");
```
Comment (richardstartin) [inline dd-java-agent/instrumentation/akka-concurrent/src/main/java/datadog/trace/instrumentation/akka/concurrent/AkkaForkJoinTaskInstrumentation.java:50]: This is incorrect - both need to be instrumented to be correct with work-stealing

File: dd-java-agent/instrumentation/akka-concurrent/src/main/java/datadog/trace/instrumentation/akka/concurrent/AkkaForkJoinExecutorTaskInstrumentation.java (hunk 1 above)
Comment (richardstartin) [reply dd-java-agent/instrumentation/akka-concurrent/src/main/java/datadog/trace/instrumentation/akka/concurrent/AkkaForkJoinExecutorTaskInstrumentation.java:46]: To make this work properly this instrumentation would need to be radically different, so no, this probably isn't long term, and only happens when Scala 2.11 is in use.

File: dd-java-agent/instrumentation/akka-concurrent/src/main/java/datadog/trace/instrumentation/akka/concurrent/AkkaWrappedRunnableInstrumentation.java (hunk 3 above)
Comment (richardstartin) [reply dd-java-agent/instrumentation/akka-concurrent/src/main/java/datadog/trace/instrumentation/akka/concurrent/AkkaWrappedRunnableInstrumentation.java:52]: This was an oversight.

File: dd-java-agent/instrumentation/akka-concurrent/src/main/java/datadog/trace/instrumentation/akka/concurrent/AkkaForkJoinExecutorTaskInstrumentation.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+
+/**
+ * akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask requires special treatment and can't
+ * be handled generically despite being a subclass of akka.dispatch.ForkJoinTask, because of its
+ * error handling.
+ *
+ * <p>This instrumentation collaborates with AkkaWrappedRunnableInstrumentation; its responsibility
```
[the remark this answers is not in the document]
Comment (richardstartin) [reply dd-java-agent/instrumentation/akka-concurrent/src/main/java/datadog/trace/instrumentation/akka/concurrent/AkkaForkJoinExecutorTaskInstrumentation.java:27]: I removed the second instrumentation and the comment.

**Pull request merged**